### ISSUE TITLE
Eliminate redundant cached information in `Node2D`: `pos`, `angle`, `_scale` and `skew`.

### ISSUE THREAD

**Issue by andre-caldas** [godotengine/godot-proposals#2742]
### Describe the project you are working on

I want to split #2738 into many smaller sub proposals. I will start with `Node2D`.

### Describe the problem or limitation you are having in your project

I will start with a bug:
```
var node = Node2D.new()
node.transform.x = Vector2(1,0)
node.transform.y = Vector2(1,1)

print("Node: ", node.transform.x, ",", node.transform.y)

node.rotate(0)
print("Node: ", node.transform.x, ",", node.transform.y)
```

The result is:
> Node: (1, 0),(1, 1)
> Node: (1, 0),(-0, 1.414214)

Notice that doing "nothing" (`node.rotate(0)`) causes `y` to change.

But this proposal is not about this bug. This bug is just a symptom. Probably, for the sake of helping the "editor tools", `Node2D` does lots of extra work and introduces some errors when trying to keep sync between `_mat` and the **redundant variables** `pos`, `angle`, `_scale` and `skew`.

There are also performance issues. For example, whenever "Node2D::translate" is called, instead of a simple vector addition, at least the following steps are performed:
1. One vector addition.
2. Two sines and two cosines are calculated.
3. Four multiplications and two sums.

Of course, this performance issue could be solved at `Node2D` level, making the code a little more complicated. The `rotate(0)` issue can be also solved at `Node2D` level, as many other issues. But the origin of those problems is that `Node2D` is doing the job of `Transform2D`.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Instead of redoing the job of `Transform2D`, `Node2D` should at most be a wrapper for `Transform2D` most basic operations: `rotate()`, `scale()` and `translate()`. Every other operation should be done through `Node2D::_mat` (that should be renamed, by the way).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This is the first step in a series of proposals. The global picture can be seen at #2738. If this succeeds and people are excited, we can start to refactor `Transform2D`. :-)

I can do the coding. :-)

Basically, eliminate `angle`, `pos`, `_scale` and `skew` in `Node2D`. This is very simple to do.

In the near future... maybe not in this proposal, but in a future new proposal, I would go a little further. I think people should be motivated to use `Transform2D` directly, except for `get_position()`, `rotate()`, `scale()` and `translate()`. Therefore I would change the API and eliminate most of the `get_` and `set_` methods. In GDScript, I would make `Node2D` really clean, by eliminating all `global_` properties. But I guess this is more a `CanvasItem` issue.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This is used very often, since it is the basic functionality of `Node2D`. It is a performance issue due to design choices that were made in favor of the *tools mode*. And it is also a "code correctness" issue. A KISS issue. A DRY issue. The workaround is just leaving as it is. :-)

### Is there a reason why this should be core and not an add-on in the asset library?

It is a refactoring of a basic core node: `Node2D`.

PS: I has been almost 20 years since I do not code professionally. I do not even know what "refactoring" actually means. Sorry if I am saying silly things! :-)

**Comment by reduz** [comment]
The reason Godot does things this way (and it does something similar on the 3D side) is not as much as performance but because of numerical precision.

If you have a transform matrix and you rotate it subsequently, you will break it (lose orthonormality, lose scale, etc). Because of this, the functions that can be operated successively (like rotate, translate, etc) operate on the actual decomposed values (rotation, scale, position, etc) instead of the transform. 

The transform is generated afterwards as a result of the modified elements. If you modify this transform, then the decomposed elements will be marked dirty (which is an optimization for performance) and updated eventually. As such, if you operate doing something like a rotation, it will decompose back the transform into components.

So, this is not really a bug, it is the way it works. Maybe for your specific use case it does not work as you expect, but for what most things users do (apply successive transforms) it works as intended.

I think it should be documented that your transform _may_ change behind your back if you use things like a negative scale or non orthonormality (what you put in your example).

**Comment by andre-caldas** [comment]
> The reason Godot does things this way (and it does something similar on the 3D side) is not as much as performance but because of numerical precision.

Yes! `Node2D` does the job `Transform2D` should be doing. I guess that in an ideal world, `Transform2D` should be virtual, so it could be represented using vectors or angles/skew/scale.

I know we are not trying to make the world ideal... I do not mind if you close this proposal. :-)
Shall I close it?

> I think it should be documented that your transform *may* change behind your back if you use things like a negative scale or non orthonormality (what you put in your example).

Yes! I think that by simply documenting it would be fine. Probably, as long as you don't make it *dirt* by messing around with `node.transform`, you are safe.

Non *orthogonality* is the problem (*scale* is not a problem).

**There is, however, a hidden bug**, since `Node2D` properly uses `Node2D::_update_xform_values()` when `node.transform` is used. And this `_update_xform_values` <a href="https://github.com/godotengine/godot/blob/5595f2406a743a5a8f85d7e5642efa3c5ca18702/scene/2d/node_2d.cpp#L117">**does properly** set `skew = transform.get_skew()`</a>. It is clear that `Node2D` was designed to support setting `node.transform`, even with *skew*.
